### PR TITLE
Add canUseAbilities() Lua Binding

### PR DESCRIPTION
**Pull request by The-Aerec**
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

Adds the canUseAbility lua function to utils. Separating from two current MRs to avoid conflicts

## Steps to test these changes

Try this function out!

### CODE REVIEW

**Review by zach2good — COMMENTED** [review]
This looks a whole lot like it should be a binding

**Comment by WinterSolstice8** [comment]
Those checks are very rigid for the states, given that it also ignores stuff like ranged, items, normal weaponskill state, etc. It's not a one size fits all function like it seems like it would be for the outside observer. Perhaps checking `ROAMING or ATTACK`, which are the "idle" and auto attack states.

File: scripts/globals/utils.lua [hunk 1]
```diff
@@ -571,3 +571,25 @@ function utils.lateralTranslateWithOriginRotation(origin, translation)
...
+        act == xi.act.MOBABILITY_USING or
+        act == xi.act.MOBABILITY_FINISH or
+        act == xi.act.MAGIC_START or
+        act == xi.act.MAGIC_CASTING or
+        mob:getStatusEffect(xi.effect.STUN) ~= nil or
+        mob:getStatusEffect(xi.effect.PETRIFICATION) ~= nil or
+        mob:getStatusEffect(xi.effect.TERROR) ~= nil or
+        mob:getStatusEffect(xi.effect.SLEEP_I) ~= nil or
+        mob:getStatusEffect(xi.effect.SLEEP_II) ~= nil or
+        mob:getStatusEffect(xi.effect.AMNESIA) ~= nil or
```
Comment (WinterSolstice8) [inline scripts/globals/utils.lua]: For completeness, you will also want to check `xi.effect.IMPAIRMENT` in case someone decides to call this on a player (or someone who's making a module uses Impairment on a mob for some reason)

**Review by WinterSolstice8 — CHANGES_REQUESTED** [review]

**Comment by zach2good** [comment]
I sat down to look at how this could be turned into a binding, and I think the name `canUseAbility` is kind of misleading. Most other functions we have that look like this take an argument about what you're checking against, rather than checking "am I able to act?".

Consider naming it something else. maybe `canUseAbilities`?

Still think this should be a binding: `if mob:canUseAbilities() then ...`

You can look at the guts of `uint8 CLuaBaseEntity::getCurrentAction()` for how to get started

**Comment by The-Aerec** [comment]
I'll start on converting this now.

**Comment by The-Aerec** [comment]
@zach2good - All Binded up

File: src/map/lua/lua_baseentity.cpp [hunk 2]
```diff
@@ -1517,6 +1517,36 @@ uint8 CLuaBaseEntity::getCurrentAction()
...
+ *  Purpose : Sees if a mob has perticular effects that stop it from doing thing
+ *  Example : mob:canUseAbilities()
+ *  Notes   : Can't use on NPCs
+ ************************************************************************/
+
+bool CLuaBaseEntity::canUseAbilities()
+{
+    XI_DEBUG_BREAK_IF(m_PBaseEntity->objtype == TYPE_NPC);
+
+    if (m_PBaseEntity == nullptr)
```
Comment (claywar) [inline src/map/lua/lua_baseentity.cpp]: m_PBaseEntity should never be null in these functions.  @zach2good correct me if I'm wrong, but I believe there is safety in the CLuaBaseEntity class itself.

Comment (The-Aerec) [reply src/map/lua/lua_baseentity.cpp]: Awesome i'll adjust on confirm, I just didn't want to introduce null chances.

File: src/map/lua/lua_baseentity.cpp [hunk 3]
```diff
@@ -1517,6 +1517,36 @@ uint8 CLuaBaseEntity::getCurrentAction()
...
+    {
+        return false;
+    }
+
+    auto* PEntity = static_cast<CBattleEntity*>(m_PBaseEntity);
+    bool actionNotFree = !(m_PBaseEntity->PAI->IsCurrentState<CInactiveState>() || m_PBaseEntity->PAI->IsCurrentState<CRangeState>() || m_PBaseEntity->PAI->IsCurrentState<CAttackState>());
```
Comment (WinterSolstice8) [inline src/map/lua/lua_baseentity.cpp]: This is still ignoring the other states that should prevent an action

After digging a bit it looks like this should be:

`bool actionNotFree = !PEntity->PAI->CanChangeState()`

`!PEntity->PAI->CanChangeState()` returns false during states like job abilities, weaponskills, casting, ranged attacks, etc. It will only return true if the state lets you switch out or your states are completed. It is used in `CPlayerController::Ability`, spell, etc like retail to prevent players from repeatedly using spells or JAs in too short of a window.

**Review by WinterSolstice8 — CHANGES_REQUESTED** [review]

File: src/map/lua/lua_baseentity.cpp (hunk 3 above)
Comment (The-Aerec) [reply src/map/lua/lua_baseentity.cpp]: This is checking that the mob is either: Auto attacking / Ranged Attacking / Idle.

If its those it returns false (Because we want the double negative for the return statement). if its anything but these states then it will return true making the below return says that its not available to do things.

It also worked successfully when I tested with an print check. Anytime the mob was not those three states or any of the status effects the binding was returning false.

Comment (WinterSolstice8) [reply src/map/lua/lua_baseentity.cpp]: There are more states than just those, such as the item state. The current implementation is not expandable for future states that may block actions. `CanChangeState()` is generic and queries whether or not any state is allowed to be actionable in it and does not add technical debt.

tl;dr: checking that does the same thing you want, but better

Comment (The-Aerec) [reply src/map/lua/lua_baseentity.cpp]: Thanks for the explanation~! Fixed

File: src/map/lua/lua_baseentity.cpp (hunk 2 above)
Comment (The-Aerec) [reply src/map/lua/lua_baseentity.cpp]: Winter confirmed what you said @claywar - null check removed.

**Review by WinterSolstice8 — APPROVED** [review]

File: src/map/lua/lua_baseentity.cpp [hunk 4]
```diff
@@ -1517,6 +1517,31 @@ uint8 CLuaBaseEntity::getCurrentAction()
...
+
+    return !(PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_SLEEP) ||
+             PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_IMPAIRMENT) ||
+             PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_SLEEP_II) ||
+             PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_STUN) ||
+             PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_AMNESIA) ||
+             PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_LULLABY) ||
+             PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_PETRIFICATION) ||
+             PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_TERROR) ||
+             actionNotFree);
```
Comment (zach2good) [inline src/map/lua/lua_baseentity.cpp]: Something like this: (XI_DEBUG_BREAK_IF brings down servers)

```cpp
bool CLuaBaseEntity::canUseAbilities()
{
    if (auto* PEntity = dynamic_cast<CBattleEntity*>(m_PBaseEntity))
    {
        return !(PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_SLEEP) ||
            PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_IMPAIRMENT) ||
            PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_SLEEP_II) ||
            PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_STUN) ||
            PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_AMNESIA) ||
            PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_LULLABY) ||
            PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_PETRIFICATION) ||
            PEntity->StatusEffectContainer->HasStatusEffect(EFFECT_TERROR) ||
            !(m_PBaseEntity->PAI->CanChangeState());
    }

    // Log about using the wrong entity type
    reuturn false;
}

File: src/map/lua/lua_baseentity.cpp [hunk 5]
```diff
@@ -1517,6 +1517,33 @@ uint8 CLuaBaseEntity::getCurrentAction()
...
+/************************************************************************
+ *  Function: canUseAbilities()
+ *  Purpose : Sees if a mob has perticular effects that stop it from doing thing
+ *  Example : mob:canUseAbilities()
+ *  Notes   : Can't use on NPCs
+ ************************************************************************/
+
+bool CLuaBaseEntity::canUseAbilities()
+{
+
```
Comment (zach2good) [inline src/map/lua/lua_baseentity.cpp]: Whitespace

**Review by zach2good — APPROVED** [review]
